tiledb/sm/cpp_api/arrow_io_impl.h

```diff
@@ -0,0 +1,834 @@
+  retval.elem_size = elem_size;
+
+  return retval;
+}
+
+int64_t flags_for_buffer(BufferInfo binfo) {
+  /*  TODO, use these defs from arrow_cdefs.h -- currently not applicable.
+      #define ARROW_FLAG_DICTIONARY_ORDERED 1
+      #define ARROW_FLAG_NULLABLE 2
+      #define ARROW_FLAG_MAP_KEYS_SORTED 4
+  */
+  (void)binfo;
+  return 0;
+}
+
+void ArrowExporter::export_(
+    std::string name, ArrowArray* array, ArrowSchema* schema) {
+  auto bufferinfo = this->buffer_info(name);
+  // TODO add checks?:
+  // -  .elem_num < INT64_MAX
+  // - check that the length of a var-len element does not exceed INT32_MAX
+  //   currently we export as "u" and "b" (int32 offsets) rather than
+  //   "U and "B" (int64 offsets)
+
+  if (schema == nullptr || array == nullptr) {
+    throw tiledb::TileDBError(
+        "ArrowExporter: received invalid pointer to output array or schema.");
+  }
+
+  auto arrow_fmt = tiledb_buffer_arrow_fmt(bufferinfo);
+  auto arrow_flags = flags_for_buffer(bufferinfo);
+
+  // lifetime:
+  //   - address is stored in ArrowSchema.private_data
+  //   - delete is called by lambda stored in ArrowSchema.release
+  CPPArrowSchema* cpp_schema =
+      new CPPArrowSchema(name, arrow_fmt.fmt_, "", arrow_flags, {}, {});
+
+  std::vector<void*> buffers;
+  if (bufferinfo.is_var) {
+    offsets_to_arrow(bufferinfo);  // convert in-place
+    buffers = {NULL, bufferinfo.offsets, bufferinfo.data};
+  } else {
+    cpp_schema =
+        new CPPArrowSchema(name, arrow_fmt.fmt_, "", arrow_flags, {}, {});
+    buffers = {NULL, bufferinfo.data};
+  }
+  cpp_schema->export_ptr(schema);
+
+  auto cpp_arrow_array = new CPPArrowArray(
+      (bufferinfo.is_var ? bufferinfo.offset_num : bufferinfo.elem_num),
+      0,  // null_num
+      0,
+      {},
+      {},
+      buffers);
+  cpp_arrow_array->export_ptr(array);
+}
+
+/* End TileDB Arrow IO internal implementation */
+/* ************************************************************************ */
+
+/* ************************************************************************ */
+/* Begin TileDB Arrow IO public API implementation */
+
+ArrowAdapter::ArrowAdapter(std::shared_ptr<tiledb::Query> query) {
+  importer_ = new ArrowImporter(query);
+  if (!importer_) {
+    throw tiledb::TileDBError(
+        "[TileDB-Arrow] Failed to allocate ArrowImporter!");
+  }
+  exporter_ = new ArrowExporter(query);
+  if (!exporter_) {
+    delete importer_;
+    throw tiledb::TileDBError(
+        "[TileDB-Arrow] Failed to allocate ArrowImporter!");
+  }
+}
+
+void ArrowAdapter::export_buffer(
+    const char* name, void* arrow_array, void* arrow_schema) {
+  exporter_->export_(
+      name, (ArrowArray*)arrow_array, (ArrowSchema*)arrow_schema);
+}
+
+void ArrowAdapter::import_buffer(
+    const char* name, void* arrow_array, void* arrow_schema) {
+  importer_->import_(
+      name, (ArrowArray*)arrow_array, (ArrowSchema*)arrow_schema);
+}
+
+ArrowAdapter::~ArrowAdapter() {
+  delete importer_;
```

Comment:
These may be unallocated if exceptions are thrown on L776 or L780. They should be constructed to nullptrs before L775 and then checked here if non-null before calling delete.